airbyte-integrations/connectors/destination-snowflake/src/main/java/io/airbyte/integrations/destination/snowflake/SnowflakeS3StagingSqlOperations.java

```diff
@@ -0,0 +1,209 @@
+
+import alex.mojaki.s3upload.MultiPartOutputStream;
+import alex.mojaki.s3upload.StreamTransferManager;
+import com.amazonaws.services.s3.AmazonS3;
+import com.amazonaws.services.s3.model.DeleteObjectsRequest;
+import com.amazonaws.services.s3.model.DeleteObjectsRequest.KeyVersion;
+import com.amazonaws.services.s3.model.ObjectListing;
+import io.airbyte.commons.lang.Exceptions;
+import io.airbyte.commons.string.Strings;
+import io.airbyte.db.jdbc.JdbcDatabase;
+import io.airbyte.integrations.destination.NamingConventionTransformer;
+import io.airbyte.integrations.destination.s3.S3DestinationConfig;
+import io.airbyte.integrations.destination.s3.csv.CsvSheetGenerator;
+import io.airbyte.integrations.destination.s3.csv.StagingDatabaseCsvSheetGenerator;
+import io.airbyte.integrations.destination.s3.util.S3StreamTransferManagerHelper;
+import io.airbyte.integrations.destination.staging.StagingOperations;
+import io.airbyte.protocol.models.AirbyteRecordMessage;
+import java.io.File;
+import java.io.PrintWriter;
+import java.nio.charset.StandardCharsets;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.UUID;
+import org.apache.commons.csv.CSVFormat;
+import org.apache.commons.csv.CSVPrinter;
+import org.apache.commons.lang3.NotImplementedException;
+import org.joda.time.DateTime;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class SnowflakeS3StagingSqlOperations extends SnowflakeSqlOperations implements StagingOperations {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(SnowflakeSqlOperations.class);
+  private static final int DEFAULT_UPLOAD_THREADS = 10; // The S3 cli uses 10 threads by default.
+  private static final int DEFAULT_QUEUE_CAPACITY = DEFAULT_UPLOAD_THREADS;
+  private static final int DEFAULT_PART_SIZE = 10;
+  private static final int UPLOAD_RETRY_LIMIT = 3;
+
+  private final NamingConventionTransformer nameTransformer;
+  private final S3DestinationConfig s3Config;
+  private AmazonS3 s3Client;
+
+  public SnowflakeS3StagingSqlOperations(final NamingConventionTransformer nameTransformer,
+                                         final AmazonS3 s3Client,
+                                         final S3DestinationConfig s3Config) {
+    this.nameTransformer = nameTransformer;
+    this.s3Client = s3Client;
+    this.s3Config = s3Config;
+  }
+
+  @Override
+  public String getStageName(final String namespace, final String streamName) {
+    return nameTransformer.applyDefaultCase(String.join("_",
+        nameTransformer.convertStreamName(namespace),
+        nameTransformer.convertStreamName(streamName)));
+  }
+
+  @Override
+  public String getStagingPath(final String connectionId, final String namespace, final String streamName, final DateTime writeDatetime) {
+    // see https://docs.snowflake.com/en/user-guide/data-load-considerations-stage.html
+    return nameTransformer.applyDefaultCase(String.format("%s/%s/%s/%02d/%02d/%02d/",
```

Comment:
I just realized that `connectionId` is randomized in the super class. Does it make sense to make this a UUID (even in the type signature)? If it's too large a change for this PR, let's make a follow up issue.

Reply:
> is there an advantage to writing it in this format as opposed to just a single time.now() timestamp?

Yes, snowflake documentation recommends this way of organizing path (which would fall under option 1 below) but it is possible to do like you say and have the timestamp as part of the file name for example (related to option 3 below).

> The COPY command supports several options for loading data files from a stage:

> 1. By path (internal stages) / prefix (Amazon S3 bucket). See [Organizing Data by Path](https://docs.snowflake.com/en/user-guide/data-load-considerations-stage.html#label-organizing-data-by-path) for information.
> 2. Specifying a list of specific files to load.
> 3. Using pattern matching to identify specific files by pattern.

But then the docs mentions:
> Pattern matching using a regular expression is generally the slowest of the three options for identifying/specifying data files to load from a stage; however, this option works well if you exported your files in named order from your external application and want to batch load the files in the same order.

An advantage of this path organization is that you can now re-run (manually?) parallel COPY commands and orce a re-load of all files for the entire hour, or the entire day or the entire month, or the entire year (depending on the part that you include or not in the path of the copy command).

> this way if a sync runs twice in an hour, we have no overwriting

There's no overwriting anyway because each csv file has also a UUID for the filename.

Reply:
ok, sgtm